### DISCUSSION TITLE
rtutorials - Interactive R And Statistics Tutorials In German

### DISCUSSION THREAD

**Discussion by luk-brue** [rstudio/learnr#807] [Show and tell]
Finally, the time has come to show off what we have been working on in the past two years :)
As a team of three, we have been writing [rtutorials (GitHub)](https://github.com/statistik-lehre/rtutorials), which is an introductory course to R and statistics in German, directed at people who have never programmed before. We wrote it to accompany a lecture in basic statistics for environmental and civil engineers during their bachelor, but it could work stand-alone as well. Our wish is to share this with fellow statistics departments in German-speaking countries as an open educational resource.

We have produced a total of 17 tutorials, starting with the basics and ending with simple linear regression. Some topics are illustrated by interactive Shiny Apps, for example on quantiles, sampling distributions and fitting a regression line (see below for some images). 

| Tutorial subject                  | Tutorial name        |
|-----------------------------------|----------------------|
| Introduction                      | `1a_intro`           |
| Exploring Functions and Arguments | `1b_funktionen`      |
| Scientific Process                | `1c_prozess`         |
| Vectors                           | `2a_vektoren`        |
| Indexing                          | `2b_indizierung`     |
| Data Frames                       | `2c_dataframes`      |
| Data Import                       | `3a_import`          |
| Scales and Factors                | `3b_skalen`          |
| Measures of Central Tendency      | `3c_zentraletendenz` |
| Measures of Spread                | `4a_dispersion`      |
| Visualization using `ggplot2`     | `4b_vis`             |
| The Pipe Operator                 | `4c_pipe`            |
| Data Wrangling                    | `4d_datawrangling`   |
| Sampling distributions (translation of ModernDive book chapter)              | `5a_sampling`        |
| t-tests                           | `5b_ttest`           |
| Correlations                      | `6a_korrelationen`   |
| Simple Linear Regression          | `6b_regression`      |

The package has gotten quite huge due to many used images and takes a while to download. Install instructions and how to deal with errors can be found in the README on GitHub. 

### Some Shiny Illustrations:
A great thing about learnr tutorials is that they can contain shiny apps. So with interactive quizzes, executable code chunks and shiny apps, directly at your fingertips in an RStudio panel, learnr is still a really good choice for writing educational resources in my opinion. Thanks to all the developers and maintainers who make this possible! 🎉 

**Getting your score at the end**
We have implemented a very basic evaluation of the quiz questions answered throughout the tutorial, using `learnr:::get_tutorial_state()`. The score is more of a little gadget than an actual evaluation, because all questions have retry enabled. For example, in the image it says: "4 out of 4 question answered throughout the whole tuorial have been answered correctly" after user clicked "Auswertung!" (*Evaluate*). 

![grafik](https://github.com/rstudio/learnr/assets/110557319/7fb7206e-2965-470f-bd8c-3abf2ec98149)

**Guessind the mean**
This is a simple app that should help understand that the positive and negative deviations of the mean should always cancel out. The user has to drag the slider to find the mean. Then a new set of numbers can be spawned and user can find the new mean. 
It is using a kinda brute force `while()` loop to sample integers that also have an integer mean, so no fractional numbers appear and everything remains intuitive. (I am sure there is a beautiful algorithm for that but it is good enough for small sets).

![Guessing the mean](https://github.com/rstudio/learnr/assets/110557319/2a03abae-c7e0-48c0-bd6f-75b53cf3f428)

**Quantiles**
This app showcases an empirical distribution function and quantiles.

![grafik](https://github.com/rstudio/learnr/assets/110557319/d319fbdc-a383-4e83-a99f-ddd663e8aedd)

**A very simple demonstration of variance...**
![grafik](https://github.com/rstudio/learnr/assets/110557319/0e2c02cf-d8d3-48a4-92dd-8e06cfb564ac)

**Fitting a regression line**
![grafik](https://github.com/rstudio/learnr/assets/110557319/87791b64-08ad-4822-a50f-37109fb6b952)

There are also a lot of carefullly crafted quiz questions that aim to give informative feedback on every answer option. 

### Disclaimer
Nobody of us is a statistics professor. Even though we all study or studied psychology, and therefore learned a lot of statistics, there can be errors. We took great care when writing this and had a review process in our little team of three for each tutorial, but it was never reviewed officially by the outside world. Hopefully this will change now! Feedback, questions and pull requests are very welcome! So please, read and fact check youself. 

### Contributors (until now)
@marie-kamp, leading the project
@gesagraf
@luk-brue
and from the outside world: 
@jeffreyohene with some pull requests

**Comment by luk-brue** [comment]
Oh, and I forgot: I patched together a very simple GitHub Action, which can show if all tutorials in the package render without errors. 

[![Tutorial Render Check](https://github.com/statistik-lehre/rtutorials/actions/workflows/tutorial_rendertest.yaml/badge.svg)](https://github.com/statistik-lehre/rtutorials/actions/workflows/tutorial_rendertest.yaml)
We needed this because it is very easy to introduce tiny errors which make the whole tutorial unrenderable. We had this happen sometimes and it was always our students who had to report it back to us. This is my very simple take on automating the feedback of "does it render on a different machine"? and preventing render-breaking bugs from being published. This can be really useful when maintaining a big set of tutorials. Notice that currently, it only checks renderability on windows because I deemed that as enough as most students use windows in our class, I did not want to go further down the rabbit hole of more GitHub Actions to make it run on other OS as well.

**Comment by luk-brue** [comment]
For the German Speaking Audience, there is a [template tutorial (GitHub)](https://github.com/statistik-lehre/rtutorials/tree/main/inst/tutorials/template) which demonstrates our basic use of CSS and some very general Markdown and Learnr Exercise Rules, such as chained setup chunks etc.

**Comment by cpauvert** [comment]
Wahoo! This look great! I can only acknowledge this amount of efforts put in there! I will share these tutorials within my research group and hopefully bring you some feedback!

Best,

**Reply by luk-brue** [reply]
Thank you! Yes indeed, a lot of effort went into creating this educational resource, we have been working on this over the course of the last two years. I'm looking forward to hear if your research group could profit from some of the contents. In case you have any questions, feel free to contact me via email or the discussion page of rtutorials. 
Best regards,
